decode_payload

```diff
@@ -0,0 +1,130 @@
+  echo "Pass CHECKINPUT=1 to process the inline checksums in addition to the final checksum"
+  exit 1
+fi
+
+PUBKEY="$1"
+FILE="$2"
+PARTOUT="$3"
+KERNEL="${4-}"
+
+MLEN=$(dd status=none bs=1 skip=12 count=8 if="${FILE}" | od --endian=big -An -vtu8 -w1024 | tr -d ' ')
+
+DESC=$(protoc --decode=chromeos_update_engine.DeltaArchiveManifest --proto_path "${SCRIPTFOLDER}"/src/update_engine "${SCRIPTFOLDER}"/src/update_engine/update_metadata.proto < <(tail -c +21 "${FILE}" |head -c "${MLEN}"))
+
+echo "${DESC}" >&2
+
+true > "${PARTOUT}"
+if [ "${KERNEL}" != "" ]; then
+  true > "${KERNEL}"
+fi
+
+PARALLEL=$(nproc)
+RUNNING=0
+MODE="" OUT="" CAT="cat" OFFSET=0 LENGTH=0 START=0 NUM_BLOCKS=0 DATAHASH="" HASH="" SIZE=0 FINALSIZE=0 FINALHASH="" KERNELSIZE=0 KERNELHASH="" SIGOFFSET=0 SIGSIZE=0
+while IFS= read -r LINE; do
+  LINE=$(echo "${LINE}" | sed 's/^ *//g')
+  case "${LINE}" in
+    "partition_operations {") MODE="partition_operations" DATAHASH="" OUT="${PARTOUT}" ;;
+    "type: REPLACE_BZ") CAT="bzcat" ;;
+    "type: REPLACE") CAT="cat" ;;
+    "data_offset:"*) OFFSET=$(echo "${LINE}" | cut -d : -f 2 | tr -d ' ') ;;
+    "data_length:"*) LENGTH=$(echo "${LINE}" | cut -d : -f 2 | tr -d ' ') ;;
+    "start_block:"*) START=$(echo "${LINE}" | cut -d : -f 2 | tr -d ' ') ;; # Part of dst_extents
+    "data_sha256_hash:"*) DATAHASH=$(echo "${LINE}" | cut -d '"' -f 2- | head -c-2 | sed 's/%/%%/g') ;; # Expected to be the last entry
+    "num_blocks:"*) NUM_BLOCKS=$(echo "${LINE}" | cut -d : -f 2 | tr -d ' ') ;; # Part of dst_extents
+    "block_size:"*) ;; # Comes after partition_operations, assumed to be 4096
+    "signatures_offset:"*) SIGOFFSET=$(echo "${LINE}" | cut -d : -f 2 | tr -d ' ') ;;
+    "signatures_size:"*) SIGSIZE=$(echo "${LINE}" | cut -d : -f 2 | tr -d ' ') ;;
+    "dst_extents {") ;; # Part of partition_operations or operations
+    "noop_operations {") MODE="noop_operations" DATAHASH="" OUT="" ;; # Comes after partition_operations, ignored
+    "new_partition_info {") MODE="new_partition_info" SIZE=0 HASH="" OUT="" ;;
+    "hash:"*) HASH=$(echo "${LINE}" | cut -d '"' -f 2- | head -c-2 | sed 's/%/%%/g') ;; # Part of new_partition_info
+    "size:"*) SIZE=$(echo "${LINE}" | cut -d : -f 2 | tr -d ' ') ;;
+    "new_info {") MODE="new_info" SIZE=0 HASH="" OUT="" ;;
+    "procedures {") MODE="procedures" DATAHASH="" OUT="" ;; # Note that only one kernel is supported and nothing else
+    "operations {") MODE="operations" DATAHASH="" ;; # Nested in "procedures"
+    "type: KERNEL") if [ "${KERNEL}" != "" ]; then OUT="${KERNEL}" ; fi ;;
+    "}") ;;
+    *) echo "unmatched: $LINE" >&2 ;;
+  esac
+  if [ "${OUT}" != "" ] && [ "${DATAHASH}" != "" ]; then
+    (
+      { dd status=none bs=1 skip="$((20 + MLEN + OFFSET))" count="${LENGTH}" if="${FILE}" | "${CAT}" | dd status=none bs=4096 seek="${START}" of="${OUT}" ; } || { echo "Write error" >&2 ; exit 1 ; }
```

Comment:
I have to benchmark if  something like `{ tail || true ; } | head` is faster for reading.

Reply:
Yes, way faster: `sys 1m32,151s` vs `sys 0m0,733s`